### ISSUE TITLE
Functional Api flatten layer requirement

### ISSUE THREAD

**Issue by moghadas76** [keras-team/keras#14334]
This model will error:

```
def build_base_model():
    """
        Base functional api
    """
    inputs = tf.keras.Input(shape=(28,28))
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    outputs = tf.keras.layers.Dense(10, activation='softmax')(x)
    model = keras.Model(inputs=inputs, outputs=outputs)
    model.compile(
        optimizer="sgd",
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    return model
```

However, by adding `Flatten` layer it will fixed while in any doc, it does not mentioned

```
def build_base_model():
    """
        Base functional api
    """
    inputs = tf.keras.Input(shape=(28,28))
    x = tf.keras.layers.Flatten()(inputs)
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    outputs = tf.keras.layers.Dense(10, activation='softmax')(x)
    model = keras.Model(inputs=inputs, outputs=outputs)
    model.compile(
        optimizer="sgd",
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    return model
```

**Comment by C3Suryansu** [comment]
My understanding of this is that, since Dense is used for working on 1-Dimensional matrix above, and your first code applied Dense to a 2 dimensional matrix input, it's giving an error. In the second case when you added flatten, it converted the 2D matrix into a 1D array which allowed Dense to perform the necessary operations on the same without any issues. 

Go through https://keras.io/api/layers/core_layers/dense/ for more information regarding this.

**Comment by jvishnuvardhan** [comment]
@moghadas76 Agree with @C3Suryansu .

If your input data is already flattened, you could use the following (just change shape=(784,) in the first input layer)

```
def build_base_model():
    """
        Base functional api
    """
    inputs = tf.keras.Input(shape=(784,))
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    x = tf.keras.layers.Dense(64, activation="relu")(x)
    outputs = tf.keras.layers.Dense(10, activation='softmax')(x)
    model = keras.Model(inputs=inputs, outputs=outputs)
    model.compile(
        optimizer="sgd",
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    return model
```

You can find an example here https://keras.io/guides/functional_api/

Also, the following is not enirely relevant but very useful resource on understanding different shape related terms used in ML https://stackoverflow.com/questions/44747343/keras-input-explanation-input-shape-units-batch-size-dim-etc


Please close the issue if this was resolved for you. Thanks!
